### ISSUE TITLE
Injection broken for PDF URLs in Firefox 60

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero-connectors#250]
I get this in the console after loading a PDF:

> Attempting to post a message to window with url "resource://pdf.js/web/viewer.html" and origin "resource://pdf.js^firstPartyDomain=bitcoin.org" from a system principal scope with mismatched origin "[System Principal]".

The change appears to be caused by this commit:

https://hg.mozilla.org/integration/mozilla-inbound/pushloghtml?fromchange=c8514e653b71479ba22a307f58f79f563ed80e9d&tochange=07ab807639ee42a407a9bdb0d374206c0f17678d

The associated bug is a restricted security bug and can't be viewed.

Firefox 60 is being released tomorrow, May 9th.

**Comment by adomasven** [comment]
I don't think this is within our control. We're using the basic injection mechanism provided by webext itself for top-frames, which a PDF page is and it is not working. I have found a [related bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1456485) on bugzilla (which references the same private associated bug as you did above), which breaks even mozilla's own screenshot taking extension.

I have created a bug on [bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1459833) for our issue.

**Comment by adomasven** [comment]
The only [info](https://bugzilla.mozilla.org/show_bug.cgi?id=1454760#c3) we have right now is that "This is intentional".

**Comment by dstillman** [comment]
Charming as ever.

So as a workaround, it seems like we could 1) determine when a page is a PDF (by watching requests for a Content-Type header?) and 2) trigger saving in background code. But we wouldn't be able to show the save popup, so the best we could do is, what, update the icon in some way?

And then when people complain we can tell them to use Chrome, at least until Firefox reconsiders this or provides WebExtensions with some method for displaying non-page UI.

**Comment by adomasven** [comment]
Web extensions support popovers that can be opened on button click (but not via context-menu commands, which is why we never used it for the progress UI). We could make an exception for firefox PDFs.

This is really annoying though and unnecessary additional complexity due to an upstream issue. We could add saving and change the button to a down-arrow icon for a simple start and see what mozilla comes up with. Surely this is an issue for many extensions and they will be pressured to make it work again.

**Comment by maxsu** [comment]
Zotero connector currently works with firefox/pdf.js. I no longer see @dstillman's console message.

Did a specific commit that fix that, and if so, which?

**Comment by dstillman** [comment]
The Mozilla restriction hasn't changed, but we put in a [workaround](bff082975a9ac3a7f6f94b06ff8bf52f7af546a8...3f7f38c01cc659299f58f0bee00b4b4e64e6d365) at the time that preserved the ability to save PDFs in Firefox. Since there's no injection, we can't show the normal save popup unless we put it in a popover.

**Comment by maxsu** [comment]
@dstillman thanks! Note the above link gives a "Not Found" over here - can you share a commit hash?

**Comment by dstillman** [comment]
Seems to be a bug with linking to commit ranges, but I've updated the link above.

**Comment by maxsu** [comment]
I see it now, thanks

**Comment by adomasven** [comment]
No updates as of today on Mozilla's end, or even movement in acknowledging this is an issue. The bug to check on is https://bugzilla.mozilla.org/show_bug.cgi?id=1454760#c42 (link to the latest current comment)

**Comment by adomasven** [comment]
Since Firefox has started reconsidering this, adding a screenshot of what this looks like on other browsers where it works:

![image](https://github.com/user-attachments/assets/4c04a953-08a9-40f7-a7b6-c3dff13095c8)

**Comment by Rob--W** [comment]
Would `action.openPopup` (MV3) or `browserAction.openPopup` work for you? The UI looks very similar.

Chrome already supports `action.openPopup` without user gesture, in Firefox that is not yet possible but we could consider prioritizing https://bugzilla.mozilla.org/show_bug.cgi?id=1799344 to enable the feature.

**Comment by adomasven** [comment]
It could probably work and once `openPopup` is available on all platforms we might transition to it in general to display the progress window for all saves. On the other hand we also run translators (in content scripts) and offer rich item saving for PDF pages where available for which we need content script support in PDFs. See https://github.com/zotero/zotero-connectors/issues/440
